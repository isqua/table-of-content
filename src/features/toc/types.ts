export type PageId = string
export type PageURL = string

export type PageDescriptor = {
    id: PageId
    title: string
    parentId: PageId
    url?: PageURL
    pages?: PageId[]
}

export type PagesIndex = Record<PageId, PageDescriptor | undefined>

export type SectionHighlight = 'parent' | 'child' | undefined

export type PageHighlight = 'active' | SectionHighlight

export type TableOfContent = {
    entities: {
        pages: PagesIndex,
    },
    topLevelIds: PageId[]
}

export type MenuItem = {
    id: PageId
    parentId: PageId
    url: PageURL
    title: string
    level: number
    highlight: PageHighlight
    hasChildren: boolean
    defaultOpenState: boolean
}
