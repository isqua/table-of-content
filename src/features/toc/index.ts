export { buildMenu } from './core/buildMenu'
export { getBreadCrumbs } from './core/getBreadCrumbs'
export { getCurrentPage } from './core/getCurrentPage'
export { useGetTocQuery } from './api/useGetTocQuery'
export * from './types'
