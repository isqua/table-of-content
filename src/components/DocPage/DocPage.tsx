import { useEffect } from 'react'

import { getBreadCrumbs, getCurrentPage, type TableOfContent } from '../../features/toc'
import { useCurrentPageUrl } from '../../hooks/useCurrentPageUrl'
import { Article } from '../Article'

type DocPageProps = {
    toc: TableOfContent
}

export function DocPage({ toc }: DocPageProps) {
    const currentUrl = useCurrentPageUrl()
    const page = getCurrentPage(toc, currentUrl)
    const breadcrumbs = getBreadCrumbs(toc, currentUrl)

    useEffect(() => {
        window.document.title = page.title
    }, [page.title])

    return (
        <Article breadcrumbs={breadcrumbs} page={page} />
    )
}
