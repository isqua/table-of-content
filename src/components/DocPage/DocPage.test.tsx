import { screen } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import { type TableOfContent } from '../../features/toc'
import { renderInApp } from '../../test'
import tocFlat from '../../test/fixtures/toc/flat.json'
import { DocPage } from './DocPage'

describe('components/DocPage', () => {
    it('should render an article by url', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/bar.html'

        renderInApp(<DocPage toc={toc} />, { url: currentUrl })

        expect(screen.getByRole('article')).toMatchSnapshot()
    })

    it('should set tab title based on the article title', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = '/baz.html'

        renderInApp(<DocPage toc={toc} />, { url: currentUrl })

        expect(document.title).toEqual('Baz')
    })
})
