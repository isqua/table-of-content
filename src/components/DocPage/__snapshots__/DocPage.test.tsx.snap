// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/DocPage > should render an article by url 1`] = `
<article
  class="_article_4ca611"
>
  <ol
    class="_breadcrumbs_296507"
  >
    <li
      class="_item_296507"
    >
      <a
        class="_link_296507"
        data-discover="true"
        href="/bar.html"
      >
        Bar
      </a>
    </li>
  </ol>
  <h1
    class="_title_4ca611"
  >
    Bar
  </h1>
</article>
`;
